File: api/speaking_band.py
"""IELTS speaking band helpers for evaluator portal."""

from __future__ import annotations

from typing import Any

CRITERIA_KEYS = ("fluency", "lexical", "grammar", "pronunciation")


def round_half(value: float) -> float:
    return round(value * 2) / 2


def compute_overall_band(scores: dict[str, Any]) -> float:
    """Mean of four criteria, rounded to nearest 0.5 (IELTS-style)."""
    if any(key not in scores or scores[key] is None for key in CRITERIA_KEYS):
        raise ValueError("All four criteria scores are required.")
    values = [float(scores[key]) for key in CRITERIA_KEYS]
    return round_half(sum(values) / len(values))


def normalize_criteria_scores(raw: dict[str, Any] | None) -> dict[str, float] | None:
    if not raw:
        return None
    out: dict[str, float] = {}
    for key in CRITERIA_KEYS:
        if key not in raw or raw[key] is None:
            return None
        out[key] = float(raw[key])
    return out


def ai_scores_to_criteria(ai_scores: dict[str, Any] | None) -> dict[str, float] | None:
    """Map ai_scores jsonb keys to rubric criteria (suggested defaults)."""
    if not ai_scores:
        return None
    mapped = {
        "fluency": ai_scores.get("fluency"),
        "lexical": ai_scores.get("lexical"),
        "grammar": ai_scores.get("grammar"),
        "pronunciation": ai_scores.get("pronunciation"),
    }
    if any(v is None for v in mapped.values()):
        return None
    return {k: float(v) for k, v in mapped.items()}

File: web/lib/speaking-band.ts
/** IELTS speaking criteria band math for evaluator portal. */

export type HumanCriteriaScores = {
  fluency: number;
  lexical: number;
  grammar: number;
  pronunciation: number;
};

export const CRITERIA_KEYS: (keyof HumanCriteriaScores)[] = [
  "fluency",
  "lexical",
  "grammar",
  "pronunciation",
];

export const CRITERIA_LABELS: Record<keyof HumanCriteriaScores, string> = {
  fluency: "Fluency & Coherence",
  lexical: "Lexical Resource",
  grammar: "Grammatical Range & Accuracy",
  pronunciation: "Pronunciation",
};

export const CRITERIA_DESCRIPTIONS: Record<keyof HumanCriteriaScores, string> = {
  fluency: "Flow, pace, linking of ideas",
  lexical: "Range & precision of vocabulary",
  grammar: "Variety & accuracy of structures",
  pronunciation: "Clarity, stress, and intonation",
};

export function roundHalf(value: number): number {
  return Math.round(value * 2) / 2;
}

export function computeOverallBand(scores: Partial<HumanCriteriaScores>): number | null {
  const values = CRITERIA_KEYS.map((key) => scores[key]);
  if (values.some((v) => v == null || Number.isNaN(v))) {
    return null;
  }
  const nums = values as number[];
  const mean = nums.reduce((sum, v) => sum + v, 0) / nums.length;
  return roundHalf(mean);
}

export function bandChipValues(min = 0, max = 9): number[] {
  const chips: number[] = [];
  for (let v = min; v <= max; v += 0.5) {
    chips.push(v);
  }
  return chips;
}

export function aiScoresToCriteria(
  aiScores: Record<string, unknown> | null | undefined,
): HumanCriteriaScores | null {
  if (!aiScores) return null;
  const fluency = aiScores.fluency;
  const lexical = aiScores.lexical;
  const grammar = aiScores.grammar;
  const pronunciation = aiScores.pronunciation;
  if (
    fluency == null ||
    lexical == null ||
    grammar == null ||
    pronunciation == null
  ) {
    return null;
  }
  return {
    fluency: Number(fluency),
    lexical: Number(lexical),
    grammar: Number(grammar),
    pronunciation: Number(pronunciation),
  };
}

export function defaultCriteriaFromReview(
  human: HumanCriteriaScores | null | undefined,
  aiScores: Record<string, unknown> | null | undefined,
): HumanCriteriaScores | null {
  if (human) return human;
  return aiScoresToCriteria(aiScores);
}
